File: attacks/sha256/semi-free-start/sha256.py
#!/usr/bin/python3

__base__ = "https://github.com/thomdixon/pysha2/blob/master/sha2/sha256.py"
__author__ = "Lukas Prokop"
__license__ = "MIT"

import copy
import struct
import binascii

F32 = 0xFFFFFFFF

_k = [
    0x428A2F98,
    0x71374491,
    0xB5C0FBCF,
    0xE9B5DBA5,
    0x3956C25B,
    0x59F111F1,
    0x923F82A4,
    0xAB1C5ED5,
    0xD807AA98,
    0x12835B01,
    0x243185BE,
    0x550C7DC3,
    0x72BE5D74,
    0x80DEB1FE,
    0x9BDC06A7,
    0xC19BF174,
    0xE49B69C1,
    0xEFBE4786,
    0x0FC19DC6,
    0x240CA1CC,
    0x2DE92C6F,
    0x4A7484AA,
    0x5CB0A9DC,
    0x76F988DA,
    0x983E5152,
    0xA831C66D,
    0xB00327C8,
    0xBF597FC7,
    0xC6E00BF3,
    0xD5A79147,
    0x06CA6351,
    0x14292967,
    0x27B70A85,
    0x2E1B2138,
    0x4D2C6DFC,
    0x53380D13,
    0x650A7354,
    0x766A0ABB,
    0x81C2C92E,
    0x92722C85,
    0xA2BFE8A1,
    0xA81A664B,
    0xC24B8B70,
    0xC76C51A3,
    0xD192E819,
    0xD6990624,
    0xF40E3585,
    0x106AA070,
    0x19A4C116,
    0x1E376C08,
    0x2748774C,
    0x34B0BCB5,
    0x391C0CB3,
    0x4ED8AA4A,
    0x5B9CCA4F,
    0x682E6FF3,
    0x748F82EE,
    0x78A5636F,
    0x84C87814,
    0x8CC70208,
    0x90BEFFFA,
    0xA4506CEB,
    0xBEF9A3F7,
    0xC67178F2,
]

_h = [
    0x6A09E667,
    0xBB67AE85,
    0x3C6EF372,
    0xA54FF53A,
    0x510E527F,
    0x9B05688C,
    0x1F83D9AB,
    0x5BE0CD19,
]


def _rotr(x, y):
    return ((x >> y) | (x << (32 - y))) & F32


def _maj(x, y, z):
    return (x & y) ^ (x & z) ^ (y & z)


def _ch(x, y, z):
    return (x & y) ^ ((~x) & z)


class SHA256:
    _output_size = 8
    blocksize = 1
    block_size = 64
    digest_size = 32
    order = 64

    def __init__(self, m=None, order=64, h=_h):
        self._counter = 0
        self._cache = b""
        self._k = copy.deepcopy(_k)
        self._h = copy.deepcopy(h)
        self.order = order

        self.update(m)

    def _compress(self, c):
        w = [0] * 64
        w[0:16] = struct.unpack("!16L", c)

        for i in range(16, self.order):
            s0 = _rotr(w[i - 15], 7) ^ _rotr(w[i - 15], 18) ^ (w[i - 15] >> 3)
            s1 = _rotr(w[i - 2], 17) ^ _rotr(w[i - 2], 19) ^ (w[i - 2] >> 10)
            w[i] = (w[i - 16] + s0 + w[i - 7] + s1) & F32

        a, b, c, d, e, f, g, h = self._h

        for i in range(self.order):
            s0 = _rotr(a, 2) ^ _rotr(a, 13) ^ _rotr(a, 22)
            t2 = s0 + _maj(a, b, c)
            s1 = _rotr(e, 6) ^ _rotr(e, 11) ^ _rotr(e, 25)
            t1 = h + s1 + _ch(e, f, g) + self._k[i] + w[i]

            h = g
            g = f
            f = e
            e = (d + t1) & F32
            d = c
            c = b
            b = a
            a = (t1 + t2) & F32

        for i, (x, y) in enumerate(zip(self._h, [a, b, c, d, e, f, g, h])):
            self._h[i] = (x + y) & F32

    def update(self, m):
        if not m:
            return

        self._cache += m
        self._counter += len(m)

        while len(self._cache) >= 64:
            self._compress(self._cache[:64])
            self._cache = self._cache[64:]

    def digest(self):
        r = copy.deepcopy(self)
        # r.update(_pad(self._counter))
        data = [struct.pack("!L", i) for i in r._h[: self._output_size]]
        return b"".join(data)

    def hexdigest(self):
        return binascii.hexlify(self.digest()).decode("ascii")


if __name__ == "__main__":

    def check(order, msg, sig, h=None):
        msg = bytes.fromhex(msg)
        m = SHA256(msg, order) if h == None else SHA256(msg, order, h)
        # m.update(msg.encode('ascii'))
        print(m.hexdigest())
        print(m.hexdigest() == sig)

    tests = {
        # "":
        #     'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855',
        # "a":
        #     'ca978112ca1bbdcafac231b39a23dc4da786eff8147c4e72b9807785afee48bb',
        # "abc":
        #     'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad',
        # "message digest":
        #     'f7846f55cf23e14eebeab5b4e1550cad5b509e3348fbc4efa3a1413d393cb650',
        # "abcdefghijklmnopqrstuvwxyz":
        #     '71c480df93d6ae2f1efad1447c66c9525e316218cf51fc8d9ed832f2daf18b73',
        # "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789":
        #     'db4bfcbd4da0cd85a60c3c37d3fbd8805c77f15fc6b1fdfe614ee0a7c8fdb4c0',
        # ("12345678901234567890123456789012345678901234567890123456789"
        #  "012345678901234567890"):
        #     'f371bc4a311f2b009eef952dd83ca80e2b60026c8e935592d0f9c308453c813e',
        # "00baf6626abc2df808da36a518c69f09b0d2ed0a79421ccfde4f559d2e42128b":
        #     'b835e56173be2b5b7177d71bf02850dc578ac855ac60f91a108eec253bd5a543',
        # "c9edzaekfwjksbz0lewkugcjxnxlmthypeaxnok3ny8eexmzsndspzw0vt2eiprz".encode('ascii'): "7e24d5acc17297ecf1978c9642056f7c5bfc43114e74d426ce978a4e25973944",
        "633965647a61656b66776a6b73627a306c65776b7567636a786e786c6d746879706561786e6f6b336e79386565786d7a736e6473707a7730767432656970727a": "7e24d5acc17297ecf1978c9642056f7c5bfc43114e74d426ce978a4e25973944"
    }

    # for inp, out in tests.items():
    #     check(inp, out)

    h = [
        0xC993C1BC,
        0x4685E40F,
        0x41270246,
        0x3D4A8BD1,
        0x24723AF6,
        0xD700757C,
        0x12E6468F,
        0xE8BA4416,
    ]

    check(
        28,
        "b28c2f27e4fce3e2cb749ca9ccf67d4c528911f24184e0e810bd0c845207030d1c07a10cca819dc8e537f9e714a994dff0313d1bc8a153485785be5a33b22d56",
        "0f1569031514c450850318400d02954e23e883d4bbf3f1af092b29dac3e055fe",
        h,
    )
    check(
        28,
        "b28c2f27e4fce3e2cb749ca9ccf67d4c528911f24184e0e810bd0c845207030decb2f1acc6599930e537f9e714a994dff0313d1bcae15bd85785be5a33b22d56",
        "0f1569031514c450850318400d02954e23e883d4bbf3f1af092b29dac3e055fe",
        h,
    )

File: encoders/nejati-preimage/src/sha256.rs
use nejati_sat::{Formula, Word};

const ROUND_CONSTANTS: [u32; 64] = [
    0x428a_2f98,
    0x7137_4491,
    0xb5c0_fbcf,
    0xe9b5_dba5,
    0x3956_c25b,
    0x59f1_11f1,
    0x923f_82a4,
    0xab1c_5ed5,
    0xd807_aa98,
    0x1283_5b01,
    0x2431_85be,
    0x550c_7dc3,
    0x72be_5d74,
    0x80de_b1fe,
    0x9bdc_06a7,
    0xc19b_f174,
    0xe49b_69c1,
    0xefbe_4786,
    0x0fc1_9dc6,
    0x240c_a1cc,
    0x2de9_2c6f,
    0x4a74_84aa,
    0x5cb0_a9dc,
    0x76f9_88da,
    0x983e_5152,
    0xa831_c66d,
    0xb003_27c8,
    0xbf59_7fc7,
    0xc6e0_0bf3,
    0xd5a7_9147,
    0x06ca_6351,
    0x1429_2967,
    0x27b7_0a85,
    0x2e1b_2138,
    0x4d2c_6dfc,
    0x5338_0d13,
    0x650a_7354,
    0x766a_0abb,
    0x81c2_c92e,
    0x9272_2c85,
    0xa2bf_e8a1,
    0xa81a_664b,
    0xc24b_8b70,
    0xc76c_51a3,
    0xd192_e819,
    0xd699_0624,
    0xf40e_3585,
    0x106a_a070,
    0x19a4_c116,
    0x1e37_6c08,
    0x2748_774c,
    0x34b0_bcb5,
    0x391c_0cb3,
    0x4ed8_aa4a,
    0x5b9c_ca4f,
    0x682e_6ff3,
    0x748f_82ee,
    0x78a5_636f,
    0x84c8_7814,
    0x8cc7_0208,
    0x90be_fffa,
    0xa450_6ceb,
    0xbef9_a3f7,
    0xc671_78f2,
];

pub struct Sha256Encoding {
    pub formula: Formula,
    pub message: [Word; 16],
    output: [Word; 8],
}

impl Sha256Encoding {
    pub fn build(rounds: usize, use_xor_clauses: bool) -> Self {
        assert!((16..=64).contains(&rounds));
        let mut formula = Formula::new(use_xor_clauses);

        let mut schedule = Vec::with_capacity(rounds);
        for word in 0..rounds {
            schedule.push(formula.new_word(Some(format!("w{word}"))));
        }
        let message: [Word; 16] = schedule[..16]
            .try_into()
            .expect("SHA-256 always has 16 input words");
        let chain: [Word; 8] = std::array::from_fn(|_| formula.new_word(None));
        let output: [Word; 8] =
            std::array::from_fn(|word| formula.new_word(Some(format!("hash{word}"))));

        let mut a = vec![chain[3], chain[2], chain[1], chain[0]];
        let mut e = vec![chain[7], chain[6], chain[5], chain[4]];
        for _ in 0..rounds {
            a.push(formula.new_word(None));
            e.push(formula.new_word(None));
        }

        for word in 16..rounds {
            let small_sigma0 = formula.new_word(None);
            let small_sigma1 = formula.new_word(None);

            let rotate1 = formula.rotate_right(&schedule[word - 15], 7);
            let rotate2 = formula.rotate_right(&schedule[word - 15], 18);
            formula.xor2(&small_sigma0[29..], &rotate1[29..], &rotate2[29..]);
            formula.xor3(
                &small_sigma0[..29],
                &rotate1[..29],
                &rotate2[..29],
                &schedule[word - 15][3..],
            );

            let rotate1 = formula.rotate_right(&schedule[word - 2], 17);
            let rotate2 = formula.rotate_right(&schedule[word - 2], 19);
            formula.xor2(&small_sigma1[22..], &rotate1[22..], &rotate2[22..]);
            formula.xor3(
                &small_sigma1[..22],
                &rotate1[..22],
                &rotate2[..22],
                &schedule[word - 2][10..],
            );

            let expanded = schedule[word];
            formula.add4(
                &expanded,
                &schedule[word - 16],
                &small_sigma0,
                &schedule[word - 7],
                &small_sigma1,
            );
        }

        let mut constants = Vec::with_capacity(rounds);
        for value in ROUND_CONSTANTS.iter().copied().take(rounds) {
            let word = formula.new_word(None);
            formula.fixed_value(&word, value);
            constants.push(word);
        }

        for (word, value) in chain.iter().zip([
            0x6a09_e667,
            0xbb67_ae85,
            0x3c6e_f372,
            0xa54f_f53a,
            0x510e_527f,
            0x9b05_688c,
            0x1f83_d9ab,
            0x5be0_cd19,
        ]) {
            formula.fixed_value(word, value);
        }

        for round in 0..rounds {
            let sigma0 = formula.new_word(None);
            let sigma1 = formula.new_word(None);
            big_sigma0(&mut formula, &sigma0, &a[round + 3]);
            big_sigma1(&mut formula, &sigma1, &e[round + 3]);

            let choose = formula.new_word(None);
            let majority = formula.new_word(None);
            formula.choose(&choose, &e[round + 3], &e[round + 2], &e[round + 1]);
            formula.majority3(&majority, &a[round + 3], &a[round + 2], &a[round + 1]);

            let temporary = formula.new_word(None);
            formula.add5(
                &temporary,
                &e[round],
                &sigma1,
                &choose,
                &constants[round],
                &schedule[round],
            );
            formula.add2(&e[round + 4], &a[round], &temporary);
            formula.add3(&a[round + 4], &temporary, &sigma0, &majority);
        }

        formula.add2(&output[0], &chain[0], &a[rounds + 3]);
        formula.add2(&output[1], &chain[1], &a[rounds + 2]);
        formula.add2(&output[2], &chain[2], &a[rounds + 1]);
        formula.add2(&output[3], &chain[3], &a[rounds]);
        formula.add2(&output[4], &chain[4], &e[rounds + 3]);
        formula.add2(&output[5], &chain[5], &e[rounds + 2]);
        formula.add2(&output[6], &chain[6], &e[rounds + 1]);
        formula.add2(&output[7], &chain[7], &e[rounds]);

        Self {
            formula,
            message,
            output,
        }
    }

    pub fn fix_output(&mut self, target: [u32; 8]) {
        for (word, value) in self.output.iter().zip(target) {
            self.formula.fixed_value(word, value);
        }
    }

    pub fn fix_message_prefix(&mut self, message: &[u32; 16], bits: usize) {
        for bit in 0..bits {
            let word = bit / 32;
            let offset = bit % 32;
            let variable = self.message[word][offset];
            self.formula
                .add_clause(vec![if (message[word] >> offset) & 1 == 1 {
                    variable
                } else {
                    -variable
                }]);
        }
    }
}

fn big_sigma0(formula: &mut Formula, output: &Word, input: &Word) {
    let rotate1 = formula.rotate_right(input, 2);
    let rotate2 = formula.rotate_right(input, 13);
    let rotate3 = formula.rotate_right(input, 22);
    formula.xor3(output, &rotate1, &rotate2, &rotate3);
}

fn big_sigma1(formula: &mut Formula, output: &Word, input: &Word) {
    let rotate1 = formula.rotate_right(input, 6);
    let rotate2 = formula.rotate_right(input, 11);
    let rotate3 = formula.rotate_right(input, 25);
    formula.xor3(output, &rotate1, &rotate2, &rotate3);
}

pub fn compression(message: &[u32; 16], rounds: usize) -> [u32; 8] {
    let mut schedule = vec![0u32; rounds.max(16)];
    schedule[..16].copy_from_slice(message);
    for word in 16..rounds {
        let sigma0 = schedule[word - 15].rotate_right(7)
            ^ schedule[word - 15].rotate_right(18)
            ^ (schedule[word - 15] >> 3);
        let sigma1 = schedule[word - 2].rotate_right(17)
            ^ schedule[word - 2].rotate_right(19)
            ^ (schedule[word - 2] >> 10);
        schedule[word] = schedule[word - 16]
            .wrapping_add(sigma0)
            .wrapping_add(schedule[word - 7])
            .wrapping_add(sigma1);
    }

    let initial = [
        0x6a09_e667u32,
        0xbb67_ae85,
        0x3c6e_f372,
        0xa54f_f53a,
        0x510e_527f,
        0x9b05_688c,
        0x1f83_d9ab,
        0x5be0_cd19,
    ];
    let [mut a, mut b, mut c, mut d, mut e, mut f, mut g, mut h] = initial;
    for round in 0..rounds {
        let sigma1 = e.rotate_right(6) ^ e.rotate_right(11) ^ e.rotate_right(25);
        let choose = (e & f) ^ ((!e) & g);
        let temporary1 = h
            .wrapping_add(sigma1)
            .wrapping_add(choose)
            .wrapping_add(ROUND_CONSTANTS[round])
            .wrapping_add(schedule[round]);
        let sigma0 = a.rotate_right(2) ^ a.rotate_right(13) ^ a.rotate_right(22);
        let majority = (a & b) ^ (a & c) ^ (b & c);
        let temporary2 = sigma0.wrapping_add(majority);

        h = g;
        g = f;
        f = e;
        e = d.wrapping_add(temporary1);
        d = c;
        c = b;
        b = a;
        a = temporary1.wrapping_add(temporary2);
    }

    [
        initial[0].wrapping_add(a),
        initial[1].wrapping_add(b),
        initial[2].wrapping_add(c),
        initial[3].wrapping_add(d),
        initial[4].wrapping_add(e),
        initial[5].wrapping_add(f),
        initial[6].wrapping_add(g),
        initial[7].wrapping_add(h),
    ]
}
